File: 5주차/김태경/네트워크.java
import java.util.*;
class Solution {
    ArrayList<Integer>[] graph;
    int n; boolean visited[];
    int[] parent;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        this.n=n;
        visited=new boolean[n];
        graph=new ArrayList[n];
        for(int i=0;i<n;i++){
            graph[i]=new ArrayList<>();
        }
        for(int i=0;i<computers.length;i++){
            for(int j=0;j<computers[0].length;j++){
                if(computers[i][j]==1){
                    graph[i].add(j);
                }
                
                
            }
        }
        parent=new int[n];
        for(int i=0;i<n;i++){
            parent[i]=i;
        }
        for(int i=0;i<n;i++){
            visited[i]=true;
            dfs(i);
            
        }
        // for(int i=0;i<n;i++){
        //     System.out.println(parent[i]);
        // }
        
        HashSet<Integer> set=new HashSet<>();
        for(int i=0;i<n;i++){
            set.add(parent[i]);
        }
        return set.size();
    }
    
    void dfs(int num){
        
        for(int i=0;i<graph[num].size();i++){
            if(!visited[graph[num].get(i)]){
                visited[graph[num].get(i)]=true;
                //System.out.println("now="+parent[graph[num].get(i)]+" par="+parent[num]);
                parent[graph[num].get(i)]=parent[num];
                dfs(graph[num].get(i));
                
                
            }
        }
    }
}

File: 5주차/박수교/네트워크.py
ans = 0
def solution(n, computers):
    
    global ans
    def dfs(graph,n,visited):
        global ans
        if graph[n] == []:
            visited[n] = True
            return
        
        for i in graph[n]:
            if not visited[i]:
                visited[i] = True
                dfs(graph, i, visited)

    
    graph = [[] for _ in range(n+1)]
    visited = [False] * (n+1)
    
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                graph[i+1].append(j+1)
        
    for k in range(1, len(graph)):
        if visited[k] == False:
            ans += 1
            dfs(graph, k, visited)
        
    return ans
